Allow customization of Prometheus rules (node-exporter)

alexandre-allard opened the pull request:
**Component**: salt, charts

**Context**: 

**Summary**:

**Acceptance criteria**: Tests are OK

---

Closes: #2448

alexandre-allard committed:
salt: add defaults for node-exporter prometheus rules

These defaults will be consumed when rendering
the prometheus-operator chart and also be exposed
to the users through the documentation.

Refs: #2448

alexandre-allard committed:
charts: add --drop-prometheus-rules option

this option allows to pass a YAML formatted file
to drop some pre-defined Prometheus alert rules.

./charts/render.py prometheus-operator \
  --namespace metalk8s-monitoring charts/prometheus-operator.yaml \
  --drop-prometheus-rules charts/drop-prometheus-rules.yaml \
  charts/prometheus-operator/

Refs: #2448

alexandre-allard committed:
charts: drop node-exporter prometheus rules

This file is to be used with the --drop-prometheus-rules
option of the charts/render.py script.
For now, it only removes the node-exporter alert rules.

Refs: #2448

alexandre-allard committed:
salt: re-rendering of chart without node-exporter alert rules

Rendered with:
  ./charts/render.py prometheus-operator \
    charts/prometheus-operator.yaml charts/prometheus-operator/ \
    --namespace metalk8s-monitoring \
    --service-config grafana metalk8s-grafana-config \
    --service-config prometheus metalk8s-prometheus-config \
    --service-config alertmanager metalk8s-alertmanager-config \
    --drop-prometheus-rules charts/drop-prometheus-rules.yaml \
    > salt/metalk8s/addons/prometheus-operator/deployed/chart.sls

Refs: #2448

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

alexandre-allard committed:
salt: add prometheus custom rules

For now it only contains node-exporter alert rules
with customization of some thresholds.

Refs: #2448

alexandre-allard committed:
tests: add pytest-bdd scenario for prometheus customization

This test is to ensure that Prometheus CSC is working
fine and we can customize Prometheus alert rules.

This scenario changes a threshold in the Prometheus
CSC ConfigMap, then re-deploys Prometheus customized
rules through Salt and checks that the value
is correctly applied on the PrometheusRule object in
Kubernetes.
Finally, the CSC is restored and the Salt formula re-run
to go back to the original state of the PrometheusRule
object.

Refs: #2448

alexandre-allard committed:
tests: remove useless import

alexandre-allard committed:
tests: rework update_service_configuration

Rework the method to handle all CSC instead
of only Dex one.
Also allow to update non existing key.

Ebaneck commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -1,37 +1,24 @@
 from ast import literal_eval
+import json
 import yaml
+import requests
```
Minor: might want to **order** them?

Ebaneck commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -238,15 +261,35 @@ def load(self, full_csc, name, namespace):
         return csc_obj
 
 
+def _query_prometheus_api(host, route):
+    cp_ip = _get_local_grain(host, 'metalk8s:control_plane_ip')
+
+    return requests.get(
+        'https://{ip}:8443/api/prometheus/api/v1/{route}'
+        .format(ip=cp_ip, route=route),
+        verify=False,
+    )
+
```
We should refactor this particular method and the following to reduce duplication.
We have the same method in `test_monitoring.py`

We could keep them in  `utils` then call them from there just like `utils.get_dict_element`

Ebaneck reviewed (changes requested):
Nicely done but still has minors to fix!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @ebaneck

alexandre-allard committed:
tests: add a get_grain helper and use it everywhere

gdemonet commented on charts/render.py:
```diff
@@ -166,6 +166,23 @@ def replace_magic_strings(rendered_yaml):
     return result
 
 
+def remove_prometheus_rules(template, drop_rules):
+    groups = []
+
+    for group in template.get('spec', {}).get('groups', []):
+        if group.get('name') in drop_rules:
+            for rule in group.get('rules', [])[:]:
+                if rule.get('alert') in drop_rules[group['name']] \
+                       or rule.get('record') in drop_rules[group['name']]:
+                    group['rules'].remove(rule)
+        if group.get('rules'):
+            groups.append(group)
+
+    template['spec']['groups'] = groups
+
+    return template
```
Do we really need to make this a mutating method? A pure variant (which can directly handle returning `None` if there's no more rules defined) would be simpler/safer I think.

alexandre-allard replied:
Haha that's what I was doing at the beginning, but though it would be more efficient to directly edit the dict, that's why there is still this return here which is useless for the moment. Will change this.

gdemonet replied:
Efficiency really isn't a concern here :P Whereas non-mutating methods are easier to test ;)

gdemonet commented on charts/render.py:
```diff
@@ -166,6 +166,23 @@ def replace_magic_strings(rendered_yaml):
+def remove_prometheus_rules(template, drop_rules):
+    groups = []
+
+    for group in template.get('spec', {}).get('groups', []):
+        if group.get('name') in drop_rules:
+            for rule in group.get('rules', [])[:]:
+                if rule.get('alert') in drop_rules[group['name']] \
+                       or rule.get('record') in drop_rules[group['name']]:
+                    group['rules'].remove(rule)
+        if group.get('rules'):
+            groups.append(group)
```
Using many `.get()` in the various `if`/`for` statements makes it not super readable IMHO... Let's try something with more variables to make it more obvious:

```python
existing_groups = template.get('spec', {}).get('groups', [])
for group in existing_groups:
    group_rules = group.get('rules', [])
    new_rules = group_rules[:]
    to_drop = drop_rules.get(group.get('name'), [])
    if to_drop:
        for rule in group_rules:
            if any(rule.get(key) in to_drop for key in ['alert', 'record']):
                new_rules.remove(rule)
    if new_rules:
        groups.append(dict(group, rules=new_rules))
```

Of course, we could also make it work with `filter` (but the benefit isn't obvious, matter of taste):

```python
    new_rules = list(filter(
        lambda rule: not any(
            rule.get(key) in to_drop for key in ['alert', 'record']
        ),
        group_rules
    ))
```

alexandre-allard replied:
Seems readable enough for you to understand. :D
Ok, will change.

gdemonet commented on charts/render.py:
```diff
@@ -236,11 +269,16 @@ def main():
     )
     sys.stdout.write('\n')
 
+    charts = []
```
It's more a list of `manifests` or `docs` than a list of `charts` (we are reading the templated output of **one** chart).

alexandre-allard replied:
true

gdemonet commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -92,30 +78,26 @@ def update_service_configuration(
+    csc_obj = csc.get(name, namespace)
+    csc_content = csc.load(csc_obj, name, namespace)
+
+    current = csc_content
+    elements = path.split('.')
+    for element in elements[:-1]:
+        current = current.setdefault(element, {})
+    current[elements[-1]] = value
```
don't we have this defined in a helper (`utils.set_dict_element` or something) already?

alexandre-allard replied:
Yes, but it'll not work if the key does not already exist.

gdemonet replied:
Maybe it should be fixed?

alexandre-allard replied:
Maybe, didn't know if it was intended or not... but since this function is not used anywhere, I'll change it;

gdemonet commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -92,30 +78,26 @@ def update_service_configuration(
-    csc_from_yaml = yaml.safe_load(
-        DEFAULT_CSC_CONFIG_YAML.format(replicas=value)
-    )
+    csc_obj = csc.get(name, namespace)
+    csc_content = csc.load(csc_obj, name, namespace)
```
Sorry it's probably just existing code, but what are those two methods doing? Why not just have one?

alexandre-allard replied:
Yes, it's already existing code and I agree that it should be done in the same method.
I wanted to change this, but I forgot. Will do it.

gdemonet commented on tests/conftest.py:
```diff
@@ -32,6 +32,7 @@ def version(request, host):
 
+
 @pytest.fixture(scope="module")
 def hostname(host):
```
We'd need to rename this fixture, since it's the minion ID, not hostname... For later.

TeddyAndrieux replied:
Nop, this one return the real hostname, not the minion ID you have another one just bellow `nodename` to retrieve the minion ID/node name (but maybe this "hostname" one is no longer needed)

gdemonet replied:
:rofl: got tricked by the collapsed lines :facepalm: please ignore!

gdemonet commented on tests/utils.py:
```diff
@@ -180,3 +178,13 @@ def set_dict_element(data, path, value, delimiter='.'):
+        output = host.check_output(
+            'salt-call --local --out=json grains.get "{}"'.format(key)
+        )
+        grain = json.loads(output)['local']
```
Can be done outside of the `host.sudo()` context

alexandre-allard replied:
Does not seem to work, at least on vagrant box.

gdemonet replied:
Oh, strange. OK let's keep it like this and investigate later.

gdemonet commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -180,6 +186,47 @@ def restore_csc(
     )
 
 
+@then(parsers.parse(
+    "we have an alert rule '{rule_name}' in group '{group_name}' with "
+    "'{path}' equal to '{value}'"
+))
```
Would make sense to define this new parser in the same commit that uses it.

alexandre-allard replied:
Totally.

gdemonet commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -180,6 +186,47 @@ def restore_csc(
+
+    def _wait_prometheus_config_reload():
+        response = requests.get(
+            'https://{}:8443/api/prometheus/api/v1/rules'.format(
```
If you're considering alert rules only (from the parser text), shouldn't you filter this query? `/api/v1/rules?type=alert`

alexandre-allard replied:
I tried, but IIRC it does not work.

alexandre-allard committed:
tests: add a wrapper around prometheus API

Add a class wrapping the call to Prometheus
API and add a fixture around that.

gdemonet commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -180,6 +186,47 @@ def restore_csc(
+    """
+
+    def _wait_prometheus_config_reload():
+        response = requests.get(
+            'https://{}:8443/api/prometheus/api/v1/rules'.format(
+                control_plane_ip
+            ),
+            verify=False
+        )
```
Configuration reload is handled without a Pod restart (I think Prom has support for both `SIGHUP` and a `/-/reload` API, probably `prometheus-operator` handles this)? No chance of getting HTTP errors when doing this query?

alexandre-allard replied:
I don't know, maybe, didn't had issue during my tests.

gdemonet reviewed (changes requested):
Design is good (except maybe putting the "drop rules" feature in `charts/render.py`, but let's tackle this later), mostly code polishing left now :)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @ebaneck

* @gdemonet

alexandre-allard committed:
tests: merge CSC get & load methods

No need for 2 methods called separatly as
we always need to load the content after
getting it.

gdemonet commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -63,15 +54,14 @@ def check_csc_configuration(
     )
 
-    csc_obj = csc.load(csc_response, name, namespace)
     return dict(csc_obj=csc_obj)
```
Why do we return here? And why wrap it in a higher-level dict? (not your code again :/ but since you're refactoring...)

gdemonet replied:
Actually, could be smart to return the `csc_obj` as a fixture and consume it from the `when`. And change the `ClusterServiceConfiguration` class to manage a single instance of CSC (not just a client), giving it methods to handle teardown at different fixture scopes... For the future ;)

alexandre-allard replied:
Agree, just trying to change as less as possible as it's not related to the ticket linked to this PR.

gdemonet commented on tests/post/features/service_configuration.feature:
```diff
@@ -9,3 +9,14 @@ Feature: Cluster and Services Configurations
+        When we update 'metalk8s-prometheus-config' CSC in namespace 'metalk8s-monitoring' 'spec.rules.node_exporter.node_filesystem_space_filling_up.warning.hours' to '48'
+        And we apply the 'metalk8s.addons.prometheus-operator.deployed.prometheus-rules' salt state
+        Then we have an alert rule 'NodeFilesystemSpaceFillingUp' in group 'node-exporter' with 'annotations.summary' equal to 'Filesystem is predicted to run out of space within the next 48 hours.'
+        And we restore original state of 'metalk8s-prometheus-config' CSC in namespace 'metalk8s-monitoring'
+        And we apply the 'metalk8s.addons.prometheus-operator.deployed.prometheus-rules' salt state
```
Oh, sorry I didn't catch this before! It's really not a good way of doing cleanup: the `given` used to find a CSC should register teardown code for resetting to the default state. Same goes for the `then` used to trigger the update with Salt, I think. Let's keep it this way for now, as we really wish to get rid of BDD, but it's not a safe approach (if the first `then` assertion fails, you will not run other `then`s, hence will not run your teardown logic).

gdemonet commented on tests/utils.py:
```diff
@@ -180,3 +180,54 @@ def set_dict_element(data, path, value, delimiter='.'):
+
+
+class PrometheusApi():
+    def __init__(self, host, port=CONTROL_PLANE_INGRESS_PORT):
```
If we really want a default value, maybe use the default Prometheus port (9090) and not the Ingress?

gdemonet commented on tests/utils.py:
```diff
@@ -180,3 +180,54 @@ def set_dict_element(data, path, value, delimiter='.'):
+    pass
+
+
+class PrometheusApi():
```
(nit) No need for the parentheses

TeddyAndrieux replied:
I disagree to me you should inherit from `object`, this one is old style Class pretty sure pylint or flake8 say something about this one

slaperche-scality replied:
Guillaume is right, in Python 3 you don't need to inherit from object explicitly.
old-style class are a thing of the past (Python 2).

Edit: see [here](https://stackoverflow.com/a/45062077) for instance.

alexandre-allard replied:
Who are you calling old?! :older_man:

TeddyAndrieux replied:
Hmm ok you right, we use python3 in pytest

gdemonet replied:
> old-style class are a thing of the past (Python 2).

Python 2 is a thing of the past :trollface:

gdemonet commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -199,6 +184,45 @@ def restore_csc(
+        except utils.PrometheusApiError as exc:
+            pytest.fail(str(exc))
+
+        for group in response.get('data', {}).get('groups', []):
+            if group.get('name') == group_name:
+                for rule in group.get('rules', []):
+                    if rule.get('name') == rule_name:
+                        assert utils.get_dict_element(rule, path) == value
+                        break
+                else:
+                    continue
+                break
+        else:
+            pytest.fail(
+                "Alert rule {} in group {} not found".format(
+                    rule_name, group_name
+                )
+            )
```
Since you've created your own helper to manage Prom API, why not give it this logic to find a specific rule? Could be more re-usable.

gdemonet replied:
And just handle comparing values in this test

alexandre-allard replied:
Yes, why not.

gdemonet commented on tests/post/steps/test_service_configuration.py:
```diff
@@ -219,7 +243,17 @@ def get(self, name, namespace):
             )
-        return response
+
+        try:
+            csc_key = "config.yaml"
+            csc_object = yaml.safe_load(response.data[csc_key])
+        except yaml.YAMLError as exc:
+            raise Exception(
```
As a general practice, only put the expression that is expected to break within a `try` block. This makes it easier for the reader to understand what part of your code is supposed to raise the exception that you're catching with `except`.

In this case, it'd be nice to add another guard against the absence of the `config.yaml` key:

```suggestion
        csc_key = "config.yaml"
        csc_data = response.data.get(csc_key)
        if csc_data is None:
            raise Exception("Missing {} key in ConfigMap".format(csc_key))

        try:
            csc_object = yaml.safe_load(csc_data)
        except yaml.YAMLError as exc:
            raise Exception(
```

alexandre-allard replied:
Yeah... I know, just copy/pasted what was below, I should have changed it...

gdemonet reviewed (dismissed):
Some more :sweat_smile:

alexandre-allard committed:
tests: fix set_dict_element to handle non existing keys

If some keys do not exist in the path, the function
will now create empty nested dictionnaries to reach
the final key.

alexandre-allard committed:
tests: get rid of a useless fixture & params

Remove a useless fixture and replace the term
service by deployment, because here we're talking
about deployments. Probably something copy/pasted
and never replaced.
Also remove useless parameters in the function.

alexandre-allard committed:
tests: refactor CSC tests

Put all the logic inside the class.
Also ensure the teardown is done and well ordered.

Ebaneck reviewed (dismissed):
Not tested entirely!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

alexandre-allard commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# Partial merge



Apparently, new commits were added to the source branch after I had queued this
Pull Request. As a result, I have merged only **part of** it into the targetted
development branches:


* :heavy_check_mark: `development/2.6`


I will now work on the following commits:


* #45a633b3ee1fbfb916e883246ec36e043bc2d080

* #d607cac54f7635d650b3eb10a1c2660067f093cf

* #6954c5f979049cea647e5d3502d30072bd70eb9a

* #c377853d055f50591661fe2df7899f60fb6c4598


**Please avoid this situation next time.**

As a friendly reminder, you can use the `wait` command to prevent me from processing your
pull-request while your work is still in progress.

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/2448-prom-customize-rules` and the
destination branch `development/2.6`.

Please resolve the conflict on **the feature branch** (`improvement/2448-prom-customize-rules`).

```bash
 $ git fetch
 $ git checkout origin/improvement/2448-prom-customize-rules
 $ git merge origin/development/2.6
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/2448-prom-customize-rules
```







*The following options are set:* **approve**
